core/sql/optimizer/ScmCostMethod.cpp

```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+// CostMethodHbaseUpdateOrDelete::numRowsToScanWhenAllKeyColumnsHaveHistograms()
+//
+// Returns an estimate of the number of rows that will be scanned as a
+// result of applying key predicates. Assumes that histograms exist for
+// all key columns.
+// -----------------------------------------------------------------------
+#pragma nowarn(262)   // warning elimination
+CostScalar
+CostMethodHbaseUpdateOrDelete::numRowsToScanWhenAllKeyColumnsHaveHistograms(
+  IndexDescHistograms & histograms,
+  const ColumnOrderList & keyPredsByCol,
+  const CostScalar & activePartitions,
+  const IndexDesc * CIDesc
+  ) const
+{
+
+  // Determine if there are single subset predicates:
+  CollIndex singleSubsetPrefixOrder;
+  NABoolean itIsSingleSubset =
+    keyPredsByCol.getSingleSubsetOrder( singleSubsetPrefixOrder );
+
+  NABoolean thereAreSingleSubsetPreds = FALSE;
+  if ( singleSubsetPrefixOrder > 0 )
+  {
+    thereAreSingleSubsetPreds = TRUE;
+  }
+  else
+  {
+    //  singleSubsetPrefixOrder==0  means either there
+    // is an equal, an IN,  or there are no key preds in the
+    // first column.
+    // singleSubsetPrefixOrder==0 AND itIsSingleSubset
+    // means there is an EQUAL or there are no key preds
+    // in the first column, check for existance of
+    // predicates in this case:
+    if (     itIsSingleSubset // this FALSE for an IN predicate
+	 AND keyPredsByCol[0] != NULL
+       )
+    {
+      thereAreSingleSubsetPreds = TRUE;
+    }
+  }
+
+
+  CMPASSERT(NOT histograms.isEmpty());
+
+  if ( thereAreSingleSubsetPreds )
+  {
+    // ---------------------------------------------------------
+    // There are some key predicates, so apply them
+    // to the histograms and get the total rows:
+    // ---------------------------------------------------------
+
+    // Get all the key preds for the key columns up to the first
+    // key column with no key preds
+    ValueIdSet singleSubsetPrefixPredicates;
+    NABoolean conflict = FALSE;
+    for ( CollIndex i = 0; i <= singleSubsetPrefixOrder; i++ )
+    {
+
+      const ValueIdSet *predsPtr = keyPredsByCol[i];
+      CMPASSERT( predsPtr != NULL ); // it must contain preds
+      singleSubsetPrefixPredicates.insert( *predsPtr );
+
+    } // for every key col in the sing. subset prefix
+
+    // Apply those key predicates that reference key columns
+    // before the first missing key to the histograms:
+    const SelectivityHint * selHint = CIDesc->getPrimaryTableDesc()->getSelectivityHint();
+    const CardinalityHint * cardHint = CIDesc->getPrimaryTableDesc()->getCardinalityHint();
+
+    RelExpr * dummyExpr = new (STMTHEAP) RelExpr(ITM_FIRST_ITEM_OP,
+				                 NULL,
+				                 NULL,
+				                 STMTHEAP);
+
+    histograms.applyPredicates( singleSubsetPrefixPredicates, *dummyExpr, selHint, cardHint);
+
+  } // if there are key predicates
+
+  // If there is no key predicates, a full table scan will be generated.
```

Comment:
If we do not take the existence of key predicates path, we never apply the selHint or cardHint. Suggest to move the call  @3761 out of the block so that it can be called when there is no key predicate case.